packages/app/components/onboarding/hcaptcha/challenge.web.tsx

```diff
@@ -0,0 +1,157 @@
+import { Logger } from "app/lib/logger";
+import { useRudder } from "app/lib/rudderstack";
+import { MY_INFO_ENDPOINT } from "app/providers/user-provider";
+
+import { toast } from "design-system/toast";
+
+import { OnboardingStepContext } from "../onboarding-context";
+import {
+  useFinishOnboarding,
+  useValidateCaptchaWithServer,
+} from "./hcaptcha-utils";
+import { siteKey } from "./sitekey";
+
+export const Challenge = () => {
+  const { redirectUri, user } = useContext(OnboardingStepContext);
+  const { validate } = useValidateCaptchaWithServer();
+  const finishOnboarding = useFinishOnboarding();
+  const { mutate } = useSWRConfig();
+  const matchMutate = useMatchMutate();
+  const { rudder } = useRudder();
+  const [hCaptchaLoaded, setHCaptchaLoaded] = useState(
+    typeof window.hcaptcha !== "undefined"
+  );
+  const [challengeRunning, setChallengeIsRunning] = useState(false);
+  const isDark = useIsDarkMode();
+  const captchaRef = useRef<HCaptcha>(null);
+
+  // this is the callback function that is called when the
+  // hcaptcha library is loaded
+  const onLoad = () => {
+    setHCaptchaLoaded(true);
+  };
+
+  const showCaptcha = async () => {
+    // Todo: skip directly to the next step if user has already a social account
+    // connected or if the user has already completed the captcha challenge
+
+    if (
+      user?.data.profile.captcha_completed_at ||
+      user?.data.profile.has_social_login
+    ) {
+      finishOnboarding(redirectUri);
+      return;
+    }
+
+    // open the captcha if the condition above is not met
+
+    setChallengeIsRunning(true);
+
+    try {
+      const token = await captchaRef.current?.execute({ async: true });
+
+      // send the response to the server and validate it
+      const status = await validate(token?.response).catch((err) => {
+        const error = err as AxiosError;
+        if (axios.isAxiosError(error)) {
+          Logger.log(error.response?.data.error.message);
+          toast.error(error.response?.data?.error?.message);
+        } else {
+          Logger.log(err?.message);
+        }
+
+        return "failed";
+      });
+
+      // if the captcha was validated successfully, we can
+      // move on to the next step
+      if (status !== "failed") {
+        await mutate(MY_INFO_ENDPOINT);
+        await matchMutate(
+          (key) => typeof key === "string" && key.includes(USER_PROFILE_KEY)
+        );
+        rudder?.track("hCaptcha challenge success");
+
+        // finish onboarding
+        finishOnboarding(redirectUri);
+      }
+    } catch (err) {
+      toast.error(
+        "Captcha challenge failed.\nPlease try again or connect a social account."
+      );
+    } finally {
+      // this has to be called to reset the captcha once validated with the server
+      captchaRef.current?.resetCaptcha();
+      setChallengeIsRunning(false);
+    }
+  };
+
+  const onExpire = () => {
+    console.log("hCaptcha Token Expired");
```

Comment:
could be inlined, it is a no-op but we have to define it, because the lib is calling it anyway..

Reply:
Done, using a noop now defined outside the component